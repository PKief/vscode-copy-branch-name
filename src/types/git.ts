import { Uri } from 'vscode';

export type GitExtensionApi = {
  getRepository(uri: Uri): GitRepository | null;
};

export type GitExtension = {
  getAPI: (version: number) => GitExtensionApi;
};

export type GitRepository = {
  state: {
    // biome-ignore lint/style/useNamingConvention: Given by the Git API
    HEAD: {
      name: string;
    };
  };
};
