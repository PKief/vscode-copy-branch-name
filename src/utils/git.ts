import * as path from 'path';
import { extensions, window, workspace } from 'vscode';
import type { GitExtension, GitRepository } from '../types/git';

interface RepositoryPickItem {
  label: string;
  description: string;
  repository: GitRepository;
}

export const getGitRepository = async (): Promise<GitRepository | null> => {
  const gitExtension =
    extensions.getExtension<GitExtension>('vscode.git')?.exports;

  const extensionApi = gitExtension?.getAPI(1);
  if (!extensionApi) {
    return null;
  }

  // Strategy 1: Use active editor's repository (primary method)
  const activeEditor = window.activeTextEditor;
  if (activeEditor) {
    const activeFilePath = activeEditor.document.uri;
    const repo = extensionApi.getRepository(activeFilePath);
    if (repo) {
      return repo;
    }
  }

  // Strategy 2: If no active editor, check available repositories
  const repositories = extensionApi.repositories.filter(
    (repo) => repo.state.HEAD && repo.state.HEAD.name
  );

  if (repositories.length === 0) {
    return null;
  }

  // If only one repository, use it directly
  if (repositories.length === 1) {
    return repositories[0];
  }

  // Multiple repositories - show picker
  return await showRepositoryPicker(repositories);
};

const showRepositoryPicker = async (
  repositories: GitRepository[]
): Promise<GitRepository | null> => {
  const items: RepositoryPickItem[] = repositories.map((repo) => {
    const branchName = repo.state.HEAD?.name || 'unknown';
    const workspaceFolder = workspace.getWorkspaceFolder(repo.rootUri);

    let label: string;
    if (workspaceFolder) {
      label = `${workspaceFolder.name} (${branchName})`;
    } else {
      // Fallback to repository root folder name
      const repoPath = repo.rootUri.fsPath;
      const repoName = path.basename(repoPath) || 'Unknown Repository';
      label = `${repoName} (${branchName})`;
    }

    return {
      label,
      description: repo.rootUri.fsPath,
      repository: repo,
    };
  });

  const selected = await window.showQuickPick(items, {
    placeHolder: 'Select repository to copy branch name from',
    matchOnDescription: true,
  });

  return selected?.repository || null;
};
