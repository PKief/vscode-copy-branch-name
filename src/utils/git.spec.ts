import {
  type Mock,
  afterAll,
  beforeEach,
  describe,
  expect,
  it,
  jest,
  mock,
  spyOn,
} from 'bun:test';
import { type Extension, window } from 'vscode';
import type { GitExtension, GitRepository } from '../types/git';
import { getGitRepository } from './git';

describe('getGitRepository', () => {
  let getExtensionMock: Mock<
    () => Pick<Extension<GitExtension>, 'exports'> | undefined
  >;

  beforeEach(() => {
    getExtensionMock = mock(() => ({
      exports: {
        getAPI: mock(() => ({
          getRepository: mock(() => ({
            rootUri: { path: '' },
            // biome-ignore lint/style/useNamingConvention: Given by the Git API
            state: { HEAD: { name: 'main' } },
          })),
        })),
      },
    }));

    mock.module('vscode', () => ({
      extensions: {
        getExtension: getExtensionMock,
      },
      window: {
        activeTextEditor: {
          document: {
            uri: undefined,
          },
        },
      },
    }));
  });

  afterAll(() => {
    jest.clearAllMocks();
  });

  it('should return null if no git extension is found', () => {
    getExtensionMock = mock(() => undefined);
    mock.module('vscode', () => ({
      extensions: {
        getExtension: getExtensionMock,
      },
    }));

    const repo = getGitRepository();
    expect(repo).toBeNull();
    expect(getExtensionMock).toMatchSnapshot();
  });

  it('should return null if no active text editor is found', () => {
    spyOn(window, 'activeTextEditor').mockReturnValue(null as never);
    const repo = getGitRepository();
    expect(repo).toBeNull();
  });

  it('should return the repository if active text editor and git extension are found', () => {
    const expectedRepository = {
      rootUri: { path: '' },
      // biome-ignore lint/style/useNamingConvention: Given by the Git API
      state: { HEAD: { name: 'main' } },
    } as GitRepository;
    const repo = getGitRepository();
    expect(repo).toStrictEqual(expectedRepository);
  });
});
