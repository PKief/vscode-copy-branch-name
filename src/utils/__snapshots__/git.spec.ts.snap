// Bun Snapshot v1, https://goo.gl/fbAQLP

exports[`getGitRepository should return null if no git extension is found 1`] = `[class Function]`;
